### PR TITLE
feat(ci): extend embedding classification to pull requests

### LINKED ISSUE #9004: Extend embedding-based classification to PRs (area labels)
Part of #9001 (PR Lifecycle v2 epic). Depends on #9002.

## Goal

Extend the existing issue classifier (`classify.py` + `all-MiniLM-L6-v2`) to also label PRs with `area/*` tags, providing automated triage signals that PRs currently lack.

## Current State

The issue classifier runs on `issues: [opened, edited]` and assigns `area/*` labels and issue types using sentence embeddings + cosine similarity. PRs currently get zero area labels (confirmed by data — none of the 30+ recent PRs have `area/*` labels).

## Changes

- Extend `classify.py` to accept a `--pr` flag
- For PRs, the input text is: title + body + changed file paths (file paths provide strong signal that issue text lacks)
- Assign `area/*` labels using the same label descriptions and thresholds from `label-descriptions.yml`
- Trigger: call from `pr-lifecycle.js` on `/accept` (not on PR open — no point classifying PRs that might get rejected)
- Add `--output-json` mode so `pr-lifecycle.js` can consume classification scores programmatically (needed by #9005 for reviewer assignment)

## Files

- `.github/scripts/label-classification/classify.py` — add PR support and JSON output mode
- `.github/workflows/classify-issues.yml` — rename to `classify.yml`, add `workflow_call` trigger
- `.github/scripts/pr-lifecycle.js` — call classification on `/accept`

**Blocks:** #9005 (auto-assignment depends on area labels and embedding infrastructure)

### PR DESCRIPTION
**Pull request by jsenko**
Closes #9004

## Summary

PRs get zero `area/*` labels today, so the triage signal issues already have is missing entirely. This extends the existing embedding classifier (`classify.py` + `all-MiniLM-L6-v2`) to pull requests, reusing the same label descriptions and thresholds from `label-descriptions.yml` — no second config, no second model.

`classify.py` gains `--pr` alongside `--issue`, and the workflow (renamed `classify-issues.yml` → `classify.yml`) runs it when a PR opens or a draft is marked ready for review.

### Two deviations from the issue, both forced

**The trigger.** #9004 specified `/accept`. That command no longer exists — #9002 removed the triage stage, and `initNewPr` now sends every non-draft PR straight to `lifecycle/ready-for-review`. PR open *is* the triage moment now, so classification hangs off the native `pull_request_target` trigger rather than a `workflow_call` from `pr-lifecycle.js`. Drafts are ignored, consistent with the orchestrator.

**No issue types for PRs.** GitHub's issue type field does not exist on pull requests, so that stage is skipped entirely — including embedding the type descriptions, which would otherwise be computed and discarded.

### The PR text layout is load-bearing

This is the part worth reviewing carefully, because the obvious implementation silently does nothing.

`all-MiniLM-L6-v2` truncates at **256 word pieces**, while the median PR title+body in this repo is **777 tokens** (p90 1411, max 2764). File paths appended *after* the body — the natural reading of "input text is title + body + changed file paths" — are never seen by the model at all. Measured over 40 recent merged PRs, scoring whether the area label implied by the conventional-commit scope was assigned:

| Text layout | Expected label found | Avg labels/PR | PRs with no labels |
|---|---|---|---|
| title + body | 65% | 2.75 | 3/40 |
| title + body + files | 68% | 2.80 | 3/40 |
| title + files + body | 74% | 3.17 | 0/40 |
| **title + directories + body** | **76%** | **2.92** | **0/40** |
| title + directories | 50% | 2.60 | 2/40 |

Directories beat raw paths because a PR touching 30 files in one package spends 30 slots saying the same thing. Prose still matters — paths alone drop to 50%.

**Raising `max_seq_length` is not the fix.** It is a config value, not an architectural limit (the transformer has 512 positions), but going to 512 drops this layout to **62%**. The model mean-pools its token embeddings, so a wider window dilutes the concentrated directory block under PR-template boilerplate. Recorded in the README and in a comment at `classify.py` so it is not retried.

### Corrections now stick (second commit)

Classification is not one-shot — issues reclassify on `issues: edited`, PRs on every draft/ready cycle. `labels_to_add` was `selected - existing`, which has no memory of a removal, so deleting a wrong label only held until the next description edit.

`get_removed_labels()` reads GitHub's own event timeline (no new state) and `labels_to_apply()` subtracts it. The classifier remains add-only: it never removes a label you added, and after your removal it never re-adds one either.

This also matters for #10160 — a classifier that reinstates its own mistakes produces a label set that looks like human agreement but is really the model agreeing with itself, which would poison the training data that issue proposes using.

## Changes

- **`classify.py`** — `--pr`, `--output-json`, `directories_of()`, `build_text()` ordering, `capped_labels()`, `get_removed_labels()`, `labels_to_apply()`, `is_pull_request()` guard rejecting `--pr` on an issue number and vice versa, lazy `SentenceTransformer` import
- **`test_classify_unit.py`** (new) — 47 offline unit tests, no network/model/`gh` required
- **`workflows/classify.yml`** — renamed from `classify-issues.yml`; separate `classify-issue` and `classify-pr` jobs, each holding only the write permission its target needs (workflow level is now `contents: read`)
- **`actions/setup-classifier/action.yaml`** (new) — shared Python + pip/model cache setup for both jobs
- **`workflows/scripts-tests.yaml`** — runs the Python unit tests
- **`label-classification/README.md`** — PR behaviour, both measurement tables, `[capped]` marker, correction semantics

### Tuning affordance

A label could clear its threshold and still be dropped by `max_labels` with nothing in the output saying so. Now marked:

```
  >>> area/CI: 0.4368 (threshold: 0.3)
      area/maven-plugin: 0.3936 (threshold: 0.4)
      area/QE: 0.3312 (threshold: 0.2)  [capped: over threshold, but max_labels=4]
```

`area/maven-plugin` scores higher than `area/QE` but is not capped — it never cleared its own 0.40 threshold. Two failure modes, two different fixes.

## Known limitation

PR thresholds are inherited from issue tuning and **run hot** — 10 of 40 sampled PRs hit the 4-label cap. Tuning them properly needs human-assigned ground truth, and no PR in the repo carries an `area/*` label yet (that is what this PR starts producing). Documented in the README as the next step, with `max_labels` flagged as the other lever.

Relatedly, #10160 records a measured finding from this work: learning from the 854 already-labeled issues beats cosine-vs-description by **17 points top-1** on the same embeddings, while swapping in a larger embedding model gains ~1 point at best. Filed as `priority/low` with the trade-offs, not actioned here.

## Test plan

- [x] 47 Python unit tests (`test_classify_unit.py`), wired into `scripts-tests.yaml`
- [x] 96 existing Node script tests still pass (no regression)
- [x] Live `--dry-run` against real PRs and issues, including the `--pr`-on-an-issue guard and the changed-files-404 degradation path
- [x] Both workflow files and the composite action parse as YAML
- [x] Issue classification path verified unchanged (same labels, issue type still assigned)
- [ ] CI: `Scripts Tests` green on this PR
- [ ] Post-merge: confirm the first real PR gets sensible `area/*` labels

Not applicable: no Java touched, so no `./mvnw` build, checkstyle, or storage-variant testing.

## Checklist

- [x] Linked issue has maintainer approval (#9004 is a maintainer-authored sub-issue of epic #9001)
- [x] No overlapping open PRs for the same issue
- [x] Checked the [Tried & Rejected list](https://github.com/Apicurio/apicurio-registry/discussions/8364)
- [x] Tests added for new code paths
- [ ] Tested locally: `./mvnw test -pl <module> -am -Dlocal` — N/A, no Java
- [ ] `./mvnw checkstyle:check -pl <module>` passes — N/A, no Java
- [x] All commits have DCO sign-off

### CODE REVIEW

File: .github/actions/setup-classifier/action.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    # model. Keep the two in sync.
+    - name: Cache model
+      uses: actions/cache@55cc8345863c7cc4c66a329aec7e433d2d1c52a9 # v6
+      with:
+        path: ~/.cache/huggingface
+        key: classify-model-all-MiniLM-L6-v2
+
+    - name: Install dependencies
+      shell: bash
+      run: pip install pyyaml sentence-transformers
```
[the remark this answers is not in the document]
Comment (jsenko) [reply .github/actions/setup-classifier/action.yaml]: Agreed, applied in bdfbded.

No actual bug here — `actions/setup-python` prepends its interpreter directory to `PATH`, so `pip`, `python` and `python3` all resolve to the configured 3.14.7 in this job. But `python -m pip` pins the install to the interpreter rather than to `PATH` resolution, which is the more robust form, and it is free.

File: .github/workflows/scripts-tests.yaml [hunk 2]
```diff
@@ -40,10 +40,27 @@ jobs:
...
+        uses: actions/setup-python@5fda3b95a4ea91299a34e894583c3862153e4b97 # v7
+        with:
+          python-version: '3.14.7'
+
+      # classify.py's unit tests need only these two. sentence-transformers is
+      # deliberately not installed: classify.py imports it lazily, so the
+      # decision logic is testable without a multi-hundred-MB torch download on
+      # every PR that touches a script.
+      - name: Install Python dependencies
+        run: pip install pyyaml numpy
```
[the remark this answers is not in the document]
Comment (jsenko) [reply .github/workflows/scripts-tests.yaml]: Applied in bdfbded, and this one caught a real inconsistency rather than just a style preference.

This branch had `classify.yml` invoking `python` while `scripts-tests.yaml` used `pip` + `python3`, in a job that configures the interpreter via `setup-python`. Both files now use `python -m pip` and `python`.

Note the wider repo convention is `python3` (see `verify.yaml`, `operator.yaml`), but those workflows have no `setup-python` step at all and deliberately use the runner system Python. Where `setup-python` *is* used, matching it is the right call.

File: .github/scripts/label-classification/classify.py [hunk 3]
```diff
@@ -134,10 +275,11 @@ def classify_issue_type(issue_embedding, type_embeddings, config):
...
-def apply_labels(repo, number, labels):
+def apply_labels(repo, number, labels, is_pr=False):
+    subcommand = "pr" if is_pr else "issue"
     for label in labels:
         subprocess.run(
-            ["gh", "issue", "edit", str(number),
+            ["gh", subcommand, "edit", str(number),
              "--repo", repo, "--add-label", label],
             check=True,
         )
```
[the remark this answers is not in the document]
Comment (jsenko) [reply .github/scripts/label-classification/classify.py:294]: Correct, applied in bdfbded.

Worth spelling out why it is a real effect and not theoretical: `labels` is `labels_to_add`, a set of `str`, and CPython randomises string hashing per process unless `PYTHONHASHSEED` is fixed. So iteration order genuinely differs between runs, and two otherwise identical runs produced logs that would not diff cleanly. No functional change — each label is an independent `gh` call. Covered by a new test asserting sorted application order.

File: .github/scripts/label-classification/classify.py [hunk 4]
```diff
@@ -1,39 +1,162 @@
...
+        capture_output=True, text=True,
+    )
+    if result.returncode != 0:
+        print(f"Warning: could not list changed files: {result.stderr.strip()}")
+        return []
+    return [line for line in result.stdout.splitlines() if line]
+
+
+def get_removed_labels(repo, number):
+    """Every label that has been taken off this issue or PR at any point.
```
[the remark this answers is not in the document]
Comment (jsenko) [reply .github/scripts/label-classification/classify.py]: Applied in bdfbded — `get_removed_labels` is now `get_removed_area_labels`, with the filter in the jq.

Accepting the change, but correcting one part of the rationale, because the docstring now has to be accurate about what this does and does not buy:

- **"make intent explicit"** — agreed, and the rename matters as much as the filter.
- **"reduce data volume"** — agreed and real. An active PR timeline is dominated by the lifecycle orchestrator churning `lifecycle/*`; PR #10086 has five such removals and zero `area/*` ones.
- **"lower the risk of future automation suppressing labels unexpectedly due to unrelated removals"** — this one does not hold. Unrelated removals were already inert: the result is subtracted from the classifier's picks, which are always `area/*`, so a `lifecycle/*` entry could never suppress anything. And the risk actually worth worrying about is the opposite one — a future automation removing an `area/*` label — which this filter does not address at all, since such a removal passes straight through it. Guarding that needs actor filtering.

Behaviour is therefore unchanged; the gain is noise and readability. The docstring now states that explicitly so the actor gap is not mistaken for solved.

File: .github/scripts/label-classification/classify.py [hunk 5]
```diff
@@ -1,39 +1,162 @@
...
+        ["gh", "api", "--paginate", f"repos/{repo}/issues/{number}/events",
+         "--jq", '.[] | select(.event == "unlabeled") | .label.name'],
+        capture_output=True, text=True,
+    )
+    if result.returncode != 0:
+        # Degrade towards the old behaviour rather than towards applying
+        # nothing: a missing history should not silently stop classification.
+        print(f"Warning: could not read label history: {result.stderr.strip()}")
+        return set()
+    return {line for line in result.stdout.splitlines() if line}
```
[the remark this answers is not in the document]
Comment (jsenko) [reply .github/scripts/label-classification/classify.py:139]: Duplicate of the comment on line 106 — addressed there, and applied in bdfbded.

File: .github/workflows/scripts-tests.yaml [hunk 6]
```diff
@@ -40,10 +40,27 @@ jobs:
...
+      - name: Run Node unit tests
         run: |
           node --test .github/scripts/pr-validation.test.js \
                       .github/scripts/pr-lifecycle.test.js \
                       .github/scripts/issue-assignment.test.js
+
+      # test_classify.py (no _unit suffix) is deliberately not run here — it is
+      # an accuracy harness that needs the model, the network and a `gh` token.
+      - name: Run Python unit tests
+        run: python3 .github/scripts/label-classification/test_classify_unit.py
```
[the remark this answers is not in the document]
Comment (jsenko) [reply .github/workflows/scripts-tests.yaml]: Duplicate of the comment on line 53 — addressed there. Both the `pip install` and the `python3` invocation in this file are changed in bdfbded.
